The cuda export is no longer the untested corner. I built the branch on a machine with CUDA 12.4 and it turned up a real defect, which is now fixed.

### The audit caught a broken export

Configuring with `-DWITH_CUDA=ON` failed immediately:

```
CMake Error at .cmake/pyre_exports.cmake:168 (message):
  the exported target 'cuda' names the imported target 'Threads::Threads',
  but nothing taught the {pyre} package configuration how to find it
```

`${CUDA_LIBRARIES}`, as the deprecated `FindCUDA` assembles it, is not the list of plain library paths it looks like — it contains the `Threads::Threads` imported target. Exporting `pyre::cuda` without accounting for it would have handed every downstream CUDA consumer an unresolved target, the same failure this PR exists to remove. The interface, as written into `pyre-targets.cmake`, reads:

```
INTERFACE_LINK_LIBRARIES "pyre::pyre;/usr/lib/.../libcudart_static.a;Threads::Threads;dl;/usr/lib/.../librt.a"
```

Fixed by recording it alongside the link, as an add-on requirement so it is resolved only for consumers who ask for the component:

```cmake
pyre_exportOptionalRequirement(cuda Threads Threads::Threads)
```

Worth saying plainly: this is the case the configure-time audit was written for. No CI leg builds with CUDA, and no consumer test would have caught it either, because nothing in the matrix would have produced the target in the first place. The check found it on the first machine that could.

### Two cosmetic fixes to the generated file

Seeing real output with five exported targets surfaced two blemishes. A package taking no extra arguments rendered as `find_dependency(Threads )`, with a stray space; it now renders cleanly. And a dependency that every consumer already resolves was being searched a second time behind a component test — visible on that machine because cmake picked the *parallel* HDF5, which puts MPI in the core interface, while the `mpi` add-on asked for it again. Add-on searches are now skipped when the core already covers them. The suppression is deliberately narrow: it only skips what the *unconditional* section resolves, so two add-ons that happen to share a package each keep their own search.

### The cuda add-on is now exercised

The mock project gained `cuda.cc`, behind `PYRE_CONSUMER_EXPECT_CUDA`, in the same shape as `mpi.cc`. It includes `<pyre/cuda.h>` — which reaches both the cuda runtime headers and our own `<pyre/memory.h>` and `<pyre/journal.h>` — and links nothing but `pyre::cuda`. That is the check that both halves of the earlier commit were right: the include directories, and the `pyre` link the target previously lacked. It touches no driver api, so it passes on a machine with no gpu attached.

### Results

On a box with CUDA 12.4, parallel HDF5, MPI, PostgreSQL 17.10, gcc 15.2 and cmake 3.31.6, all five targets export — `pyre::pyre`, `pyre::journal`, `pyre::cuda`, `pyre::mpi`, `pyre::postgres` — and the configuration announces all five components. The generated searches come out as:

```cmake
find_dependency(HDF5 COMPONENTS C HL)
find_dependency(MPI COMPONENTS CXX)
if("cuda" IN_LIST pyre_FIND_COMPONENTS)     ...
if("postgres" IN_LIST pyre_FIND_COMPONENTS) ...
```

HDF5 and MPI unconditional, correctly, because that build resolved a parallel HDF5 and MPI is therefore in `libpyre`'s own interface; cuda and postgres behind their components, so a consumer wanting neither is asked for neither. The full consumer suite passes there — journal, pyre, mpi, cuda and h5.

That build also exercises the postgres export for the first time, since PostgreSQL is present on that machine and absent from every CI runner.

Locally, where HDF5 is serial, the same branch still produces a configuration with an unconditional HDF5 search and MPI behind its component — the narrow suppression behaves as intended — and the four applicable consumer drivers pass.

Remaining from my earlier list: the absolute `CMAKE_INSTALL_RPATH`, and the parallel-versus-serial HDF5 flavour mismatch. Also still open, and now better understood: the cuda interface carries absolute paths from `FindCUDA`, so a cuda-enabled install is not relocatable. Migrating to `CUDAToolkit` and its imported targets would fix that, but it is a change with its own risk and belongs on its own.

--- a/.cmake/pyre-config.cmake.in
+++ b/.cmake/pyre-config.cmake.in
@@ -12,8 +12,12 @@ include(CMakeFindDependencyMacro)
 # are generated from the places where the build established those links, so this list is exactly
 # what this particular installation needs, no more and no less
 @PYRE_EXPORT_REQUIREMENTS@
-# our headers are c++23, strictly: the exported targets carry the standard as a usage
+# our headers are c++23, strictly: the exported targets carry the standard itself as a usage
 # requirement, and this turns off the compiler extensions that would otherwise ride along
+# N.B.: cmake has no way to make this part of a usage requirement, because {CXX_EXTENSIONS} is
+# a plain target property rather than an interface one; a target only picks the setting up when
+# it is created, so ask for {pyre} before you declare the targets that will link against it, or
+# they will quietly compile as {gnu++23}
 set(CMAKE_CXX_EXTENSIONS OFF)
 
 # load the exported targets

--- a/.cmake/pyre_exports.cmake
+++ b/.cmake/pyre_exports.cmake
@@ -19,6 +19,18 @@ function(pyre_exportTarget target component)
 endfunction(pyre_exportTarget)
 
 
+# record cmake code that the generated configuration must run before any of the searches
+# use it for whatever has to be true for a search to behave, such as a language the find module
+# needs or a hint that steers it towards the right build of a package
+function(pyre_exportPrologue text)
+  # make sure we are still allowed to record
+  pyre_exportGate("prologue")
+  # and stash it; the text is kept whole, newlines and all
+  set_property(GLOBAL APPEND PROPERTY PYRE_EXPORT_PROLOGUE "${text}")
+  # all done
+endfunction(pyre_exportPrologue)
+
+
 # record an external package that one of our core targets names in its link interface
 # {package} is what to look for, {target} is the imported target the search must produce, and
 # the remaining arguments are handed to {find_dependency} verbatim
@@ -99,6 +111,7 @@ endfunction(pyre_exportSearch)
 # recorded requirements as cmake code, and hand the text back through {variable}
 function(pyre_exportRequirements variable)
   # retrieve what the build recorded
+  get_property(prologue GLOBAL PROPERTY PYRE_EXPORT_PROLOGUE)
   get_property(targets GLOBAL PROPERTY PYRE_EXPORT_TARGETS)
   get_property(components GLOBAL PROPERTY PYRE_EXPORT_COMPONENTS)
   get_property(requirements GLOBAL PROPERTY PYRE_EXPORT_REQUIREMENTS)
@@ -109,6 +122,11 @@ function(pyre_exportRequirements variable)
   set(known "")
   set(core "")
   set(text "")
+
+  # whatever has to be in place before we go looking for anything
+  foreach(chunk IN LISTS prologue)
+    string(APPEND text "${chunk}\n")
+  endforeach()
 
   # tell consumers which pieces this installation carries, so that {check_required_components}
   # can answer {find_package(pyre COMPONENTS ...)} correctly

--- a/.cmake/pyre_h5.cmake
+++ b/.cmake/pyre_h5.cmake
@@ -60,6 +60,19 @@ function(pyre_h5Lib)
     target_link_libraries(pyre HDF5::HDF5)
     # which puts it in our exported link interface, so downstream projects have to be able to
     # resolve it before they load our targets; teach the package configuration to do it for them
+    # {FindHDF5} probes the hdf5 c toolchain with its {h5cc} wrapper and needs a c compiler to
+    # do it; a consumer whose own project is c++ only would otherwise be told, from deep inside
+    # the module, that the wrapper cannot compile a minimal hdf5 program
+    pyre_exportPrologue(
+"if(NOT CMAKE_C_COMPILER_LOADED)
+  enable_language(C)
+endif()")
+    # steer them towards the same flavour of hdf5 we linked; without this a consumer of a
+    # parallel build resolves whichever hdf5 the module prefers, and a serial one against our
+    # parallel headers is a mismatch they have to diagnose themselves
+    if(HDF5_IS_PARALLEL)
+      pyre_exportPrologue("set(HDF5_PREFER_PARALLEL ON)")
+    endif()
     pyre_exportRequirement(HDF5 HDF5::HDF5 "COMPONENTS C HL")
     # a parallel build of hdf5 exposes {mpi.h} through its public headers, so anything that
     # includes them needs the mpi usage requirements; the plain signature propagates them to

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -102,6 +102,9 @@ if(MPI_FOUND)
     # leave a marker
     message(STATUS "We have MPI; ask for parallel HDF5: ${HDF5_PREFER_PARALLEL}")
 endif()
+# our own sources are c++ only, so nothing has enabled the c language yet; {FindHDF5} probes
+# the hdf5 c toolchain with its {h5cc} wrapper and cannot do it without a c compiler
+enable_language(C)
 # hunt it down; pyre::h5 wraps the hdf5 C api directly, so the C++ component is no longer needed
 find_package(HDF5 COMPONENTS C HL)
 # report

--- a/etc/cmake/consumer/CMakeLists.txt
+++ b/etc/cmake/consumer/CMakeLists.txt
@@ -15,6 +15,7 @@ project(pyre-consumer LANGUAGES CXX)
 option(PYRE_CONSUMER_EXPECT_HDF5 "the installation was built against hdf5" OFF)
 option(PYRE_CONSUMER_EXPECT_MPI "the installation carries the mpi add-on" OFF)
 option(PYRE_CONSUMER_EXPECT_CUDA "the installation carries the cuda add-on" OFF)
+option(PYRE_CONSUMER_EXPECT_POSTGRES "the installation carries the postgres add-on" OFF)
 
 # deliberately, there is no {find_package(HDF5)} or {find_package(MPI)} ahead of this call, and
 # no {CMAKE_CXX_STANDARD} anywhere in this file: resolving the dependencies of the exported
@@ -122,6 +123,22 @@ if(PYRE_CONSUMER_EXPECT_MPI)
   target_link_libraries(mpi PRIVATE pyre::mpi)
   # and it must run cleanly
   add_test(NAME mpi COMMAND mpi)
+endif()
+
+# the postgres add-on, on the same terms as mpi above
+if(PYRE_CONSUMER_EXPECT_POSTGRES)
+  # asking for the component is what sends the configuration looking for libpq
+  find_package(pyre REQUIRED COMPONENTS postgres)
+  # which must have resolved the imported target the add-on names
+  if(NOT TARGET PostgreSQL::PostgreSQL)
+    message(FATAL_ERROR "asking for the 'postgres' component did not locate libpq")
+  endif()
+  # compiling this proves the libpq headers arrive through {pyre::postgres}
+  add_executable(postgres postgres.cc)
+  # the one link a client should need
+  target_link_libraries(postgres PRIVATE pyre::postgres)
+  # and it must run cleanly
+  add_test(NAME postgres COMMAND postgres)
 endif()
 
 # the cuda add-on, on the same terms as mpi above

--- a/etc/cmake/consumer/CMakeLists.txt
+++ b/etc/cmake/consumer/CMakeLists.txt
@@ -20,6 +20,9 @@ option(PYRE_CONSUMER_EXPECT_CUDA "the installation carries the cuda add-on" OFF)
 # no {CMAKE_CXX_STANDARD} anywhere in this file: resolving the dependencies of the exported
 # targets and raising the language standard are the package configuration's job, and whether it
 # does them is precisely what this project tests
+# note also that this comes before any {add_executable} below, which is what lets the
+# configuration turn the compiler extensions off for our targets; a project that declares its
+# targets first would quietly get {gnu++23} instead
 find_package(pyre REQUIRED)
 
 # the export set: these are the two targets the package promises

--- a/etc/cmake/consumer/postgres.cc
+++ b/etc/cmake/consumer/postgres.cc
@@ -1,0 +1,30 @@
+// -*- C++ -*-
+// -*- coding: utf-8 -*-
+//
+// michael a.g. aïvázis <michael.aivazis@para-sim.com>
+// (c) 1998-2026 all rights reserved
+
+
+// get the postgres wrappers; this header reaches {libpq-fe.h}, so it compiles only if the
+// libpq include directories arrive through the exported {pyre::postgres} target
+#include <pyre/postgres.h>
+
+
+// verify that the postgres add-on of an installation that carries it is reachable by a client
+// that asks for the component and names nothing but {pyre::postgres}
+// there is no connection attempt here on purpose: this checks the package, not the runtime, so
+// it has to pass on a machine with no server of its own; the {pyre.db} and {pyre-postgres}
+// suites are the ones that want a live database
+int
+main()
+{
+    // name a type from the wrappers; requiring it to be complete is what forces the compiler
+    // to work through {libpq-fe.h}
+    static_assert(sizeof(pyre::postgres::connection_t) > 0);
+
+    // all done
+    return 0;
+}
+
+
+// end of file

